### PR TITLE
add launch and stop

### PR DESCRIPTION
**Pull request by xudafeng**

### CODE REVIEW

File: lib/helper.js [hunk 1]
```diff
@@ -16,11 +16,26 @@
 const macacaUtils = require('macaca-utils');
 
 var _ = macacaUtils.merge({}, macacaUtils);
+var childProcess = require('child_process');
```
Comment (xudafeng) [inline lib/helper.js]: tweak style

File: lib/macaca-android.js [hunk 2]
```diff
@@ -102,7 +103,37 @@ Android.prototype.initDevice = function *() {
     this.adb.setDeviceId(device.udid);
     this.udid = device.udid;
   } else {
-    console.log('no device');
+    console.log('no device, now create one from avd');
```
Comment (xudafeng) [inline lib/macaca-android.js]: logger.info?

File: lib/macaca-android.js [hunk 3]
```diff
@@ -102,7 +103,37 @@ Android.prototype.initDevice = function *() {
...
+      console.log('no avd created!');
+    } else {
+      data.map(avd => {
+        if(/Name:/.test(avd)) {
+          avd = _.trim(avd.split(':')[1]);
+          avdArr.push(avd);
+        }
+      });
+      _.exec(`${emulatorCommand} -avd ${avdArr[1]}`);
+      yield _.sleep(30000);
```
Comment (xudafeng) [inline lib/macaca-android.js]: why 30000 s?

File: lib/macaca-android.js [hunk 4]
```diff
@@ -102,7 +103,37 @@ Android.prototype.initDevice = function *() {
...
+      devices = yield ADB.getDevices();
+
+      if (devices.length) {
+        device = devices[0];
+        this.adb.setDeviceId(device.udid);
+        this.udid = device.udid;
+      }else {
+        console.log('emulator start failed or too slow!');
```
Comment (xudafeng) [inline lib/macaca-android.js]: logger.info

File: lib/macaca-android.js [hunk 5]
```diff
@@ -102,7 +103,37 @@ Android.prototype.initDevice = function *() {
...
+    var emulatorCommand = path.resolve(env.ANDROID_HOME, 'tools', 'emulator');
+    var androidCommand = path.resolve(env.ANDROID_HOME, 'tools', 'android');
+
+    var data = yield _.exec(`${androidCommand} list avd`);
+    data = data.split(EOL);
+    data.shift();
+    var avdArr = [];
+
+    if (data.length === 0) {
+      console.log('no avd created!');
```
Comment (xudafeng) [inline lib/macaca-android.js]: logger

File: lib/macaca-android.js [hunk 6]
```diff
@@ -56,6 +56,7 @@ Android.prototype.startDevice = function *(caps) {
 };
 
 Android.prototype.stopDevice = function *() {
+  _.exec('adb emu kill').then(() => {});
```
Comment (zenzhu) [inline lib/macaca-android.js]: return _.exec('adb emu kill').then(() => {});

File: lib/macaca-android.js [hunk 7]
```diff
@@ -56,6 +56,7 @@ Android.prototype.startDevice = function *(caps) {
 };
 
 Android.prototype.stopDevice = function *() {
```
Comment (zenzhu) [inline lib/macaca-android.js]: remote *

File: lib/macaca-android.js [hunk 8]
```diff
@@ -102,7 +103,37 @@ Android.prototype.initDevice = function *() {
...
+
+    var data = yield _.exec(`${androidCommand} list avd`);
+    data = data.split(EOL);
+    data.shift();
+    var avdArr = [];
+
+    if (data.length === 0) {
+      console.log('no avd created!');
+    } else {
+      data.map(avd => {
```
Comment (zenzhu) [inline lib/macaca-android.js]: var avdArr = data.filter(avd => /Name:/.test(avd)).map(avd => _.trim(avd.split(':')[1]))

File: lib/macaca-android.js [hunk 9]
```diff
@@ -102,7 +103,37 @@ Android.prototype.initDevice = function *() {
...
+        if(/Name:/.test(avd)) {
+          avd = _.trim(avd.split(':')[1]);
+          avdArr.push(avd);
+        }
+      });
+      _.exec(`${emulatorCommand} -avd ${avdArr[1]}`);
```
Comment (zenzhu) [inline lib/macaca-android.js]: yield _.exec(`${emulatorCommand} -avd ${avdArr[1]}`); ?

**Pull request merged**